sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/binaryfile/BinaryFileDataSource.scala

```diff
@@ -0,0 +1,83 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.binaryfile
+
+import org.apache.spark.annotation.{Experimental, Since}
+import org.apache.spark.sql.types._
+
+/**
+ * This "binaryFile" data source format implements Spark SQL data source API for loading binary
+ * file data as `DataFrame`.
+ *
+ * The loaded `DataFrame` has two columns, the schema is:
+ *  - content: `BinaryType` (binary data of the file content)
+ *  - status: `StructType` (the file status information)
+ *
+ * The schema of "status" column described above is:
+ *  - path: `StringType` (the file path)
+ *  - modificationTime: `TimestampType` (last modification time of the file, on some FS
+ *                                       implementation, this might be not available
+ *                                       and fallback to some default value.)
+ *  - length: `LongType` (the file length)
+ *
+ * To use binary file data source, you need to set "binaryFile" as the format in `DataFrameReader`
+ * and optionally specify the data source options, available options include:
+ *  - pathGlobFilter: Only include files with path matching the glob pattern.
+ *                    The glob pattern keeps the same behavior with Hadoop API
+ *                    `org.apache.hadoop.fs.FileSystem.globStatus(pathPattern)`
+ *
+ * In order to control the partition size, we can set spark sql configuration
+ * `spark.sql.files.maxPartitionBytes` and `spark.sql.files.openCostInBytes`.
+ *
+ * Example:
+ * {{{
+ *   // Scala
+ *   val df = spark.read.format("binaryFile")
+ *     .option("pathGlobFilter", "*.txt")
+ *     .load("path/to/fileDir")
+ *
+ *   // Java
+ *   Dataset<Row> df = spark.read().format("binaryFile")
+ *     .option("pathGlobFilter", "*.txt")
+ *     .load("path/to/fileDir");
+ * }}}
+ *
+ * @note This binary file data source does not support saving dataframe to binary files.
+ * @note This class is public for documentation purpose. Please don't use this class directly.
+ * Rather, use the data source API as illustrated above.
+ */
+@Experimental
+@Since("3.0.0")
+class BinaryFileDataSource private() {}
+
+object BinaryFileDataSource {
+
+  private val fileStatusSchema = StructType(
+    StructField("path", StringType, false) ::
+    StructField("modificationTime", TimestampType, false) ::
+    StructField("length", LongType, false) :: Nil)
+
+  /**
+   * The schema of the dataframe returned by binaryFile data source.
+   * See doc in `BinaryFileDataSource`
+   */
+  val binaryFileSchema = StructType(
+    StructField("content", BinaryType, false)::
```

Comment:
* space before `::`
* just a note: we might keep this column nullable in case to handle potential I/O failures